### ISSUE TITLE
-inf scores when generating with do_sample

### ISSUE THREAD

**Issue by fColangelo** [huggingface/transformers#28906]
### System Info

- `transformers` version: 4.37.1
- Platform: Linux-5.4.0-169-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.0
- Accelerate version: 0.24.1
- Accelerate config:    not found
- PyTorch version (GPU?): 1.14.0a0+44dac51 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

To reproduce run the following script:
```
from transformers import AutoModelForCausalLM, AutoTokenizer
"""qconf = BitsAndBytesConfig {
  "bnb_4bit_compute_dtype": "bfloat16",
  "bnb_4bit_quant_type": "nf4",
  "bnb_4bit_use_double_quant": false,
  "llm_int8_enable_fp32_cpu_offload": false,
  "llm_int8_has_fp16_weight": false,
  "llm_int8_skip_modules": null,
  "llm_int8_threshold": 6.0,
  "load_in_4bit": true,
  "load_in_8bit": false,
  "quant_method": "bitsandbytes"
}"""

model = AutoModelForCausalLM.from_pretrained("mistralai/Mixtral-8x7B-Instruct-v0.1", quantization_config=qconf)
tokenizer = AutoTokenizer.from_pretrained("mistralai/Mixtral-8x7B-Instruct-v0.1")

input_ids = tokenizer.encode("<s> [INST] Howdy! [/INST]", return_tensors="pt")
out = model.generate(tokenizer.encode("<s> [INST] Howdy! [/INST]", return_tensors="pt"), 
                     max_length=4000, 
                     pad_token_id = tokenizer.eos_token_id,
                     return_dict_in_generate=True,
                     output_scores =True,
                     temperature=0.8,
                     top_k= 50,
                     do_sample= True)
print(out.scores)
```
out.scores will have almost only -inf values. The problem disappears when do_sample is removed.

### Expected behavior

logits are returned instead of -inf (happens when setting do_sample=False)

**Comment by gante** [comment]
Hi @fColangelo 👋 

That happens because `top_k` sets the scores of all but the `top_k` most likely tokens to `-inf` (50 in your case). If you want the model logits, have a look at this feature that is about to get merged: https://github.com/huggingface/transformers/pull/28667

**Comment by fColangelo** [comment]
Hi and thanks for the answer! I noticed this since I was trying to calculate a perplexity for an answer. I guess my code was incorrect, since the probability values for the selected tokens should not be -inf. I will check, confirm and close, ty!

**Comment by fColangelo** [comment]
It was in fact a bug in my indexing. Thanks for the info!
